src/lseq/ident.rs

```diff
@@ -0,0 +1,314 @@
+use serde::{Deserialize, Serialize};
+use bitvec::vec::*;
+
+const BOUNDARY: u64 = 10;
+
+const INITIAL_BASE: u32 = 3; // start with 2^8
+
+/// A tree identifier uniquely locates a character in an LSeq tree
+/// it represents the path that needs to be taken in order to reach
+/// the character. At each level we store the index of the child tree node
+/// as well as the id of the site that inserted that node. This resolves conflicts where
+/// two sites decide to pick the same child index to allocate a fresh node
+#[derive(Debug, PartialEq, PartialOrd, Ord, Eq, Clone, Serialize, Deserialize, Hash)]
+pub struct Identifier {
+    path: Vec<(u64, u32)>,
+    // site_id: u64,
+    // counter: u64
+}
+impl Identifier {
+    /// Get the size of an identifier
+    pub fn len(&self) -> usize {
+        self.path.len()
+    }
+}
+/// Generates fresh identifiers
+///
+/// These identifiers represent a path in an exponential tree. At each level of the tree the amount
+/// of children doubles. This means that we can store a lot of nodes in a very low height!
+///
+/// The LSeq tree also has an additional restriction: at each level the minimum and maximum nodes
+/// cannot be chosen when allocating fresh nodes. This is to ensure there is always a free node
+/// that can be used to create a lower level.
+//#[derive(Serialize, Deserialize)]
+pub struct IdentGen {
+    initial_base_bits: u32,
+    strategy_vec: BitVec,
+    /// Site id of this tree
+    pub site_id: u32,
+    // clock: u64
+}
+
+impl IdentGen {
+    /// Create a fresh tree with 0 node
+    pub fn new(site_id: u32) -> IdentGen {
+        Self::new_with_args(INITIAL_BASE, site_id)
+    }
+
+    /// Create a tree with a custom initial size
+    pub fn new_with_args(base: u32, site_id: u32) -> IdentGen {
+        IdentGen { initial_base_bits: base, strategy_vec: BitVec::new(), site_id }
+    }
+
+    /// The smallest possible node in a tree
+    pub fn lower(&self) -> Identifier {
+        Identifier { path: vec![(0, 0)] }
+    }
+
+    /// The absolute largest possible node
+    pub fn upper(&self) -> Identifier {
+        Identifier { path: vec![(2u64.pow(self.initial_base_bits) - 1, 0)] }
+    }
+    /// Allocates a new identifier between p and q.
+    /// Requires that p < q and will produce a new identifier z, p < z < q
+    ///
+    /// The way to think about this is to consider the problem of finding a short number between
+    /// two decimal numbers. For example, between 0.2 and 0.4 we'd like to return 0.3
+    /// But what about 0.2 and 0.3? Well a nice property is that if we add a digit to 0.2, the
+    /// number produced will be larger than 0.2 but smaller than 0.3.
+    ///
+    /// If you view identifiers p and q as decimal numbers 0.p and 0.q. Then all we're doing is
+    /// finding a number between them!
+    ///
+    ///
+    pub fn alloc(&mut self, p: &Identifier, q: &Identifier) -> Identifier {
+        assert!(p < q, "lower bound should be smaller than upper bound!");
```

Comment:
Just had a thought: What if we returned `Result<Identifier, Identifier>` (or `Either`) here? The `Err` side means that `p` was larger than `q`, and `Ok` means `p` was less than `q`.

Reply:
That's possible, I feel like it keeps things conceptually simpler if `p < q` and closer to the paper description of the algorithm. I'm not sure what the project's policy is wrt to function preconditions.

Reply:
Sure, that makes sense. I tend to veer in the direction of providing error handling to the user -- especially in an API. But it would be good to hear from the `rust-crdt` maintainers :)

Reply:
yea for sure! I'd be fine with wrapping this all in a `Result` for sure,

Reply:
I could do it in a follow up PR if that works, or if you don't get around to it :)

Reply:
I'll work on it as soon your PR gets merged

Reply:
I made `alloc` handle the case where `q < p`. It now only panics if `q = p` should that be wrapped in an `Option/Result`? There's nothing useful you can really do in the case of an error. In theory that could mean multiple sites start diverging their state (if one ended up erroring and trying to recover)

Reply:
Well, from an application's point of view: the useful thing is that it resulted in an `Err` or `None` case and can log this and trying to continue in some way, if possible. If you `assert` you're forcing their application to crash -- unless they do some catch, unwind logic.

At the same time, I think this might be fine if it's a rare case? I'm not quite sure really. :)

Reply:
True enough, it's just that in this case if you _ever_ get anything other than a successful response the only correct course of action is to stop the whole application.